New Reset Handler tutorial (#12)
cyberbotics/ros2_documentation: source/Tutorials/Advanced/Simulators/Webots/Simulation-Reset-Handler.rst

Prerequisites
-------------

Before proceeding with this tutorial, make sure you have completed the following:

- Understanding of ROS 2 nodes and topics covered in the beginner :doc:`../../../../Tutorials`.
- Knowledge of Webots and ROS 2 and its interface package.
- Familiarity with :doc:`./Setting-Up-Simulation-Webots-Basic`.

Reset Handler for Simple Cases (Controllers Only)
-------------------------------------------------

In the launch file of your package, add the ``respawn`` parameter.

.. code-block:: python

  def generate_launch_description():
      robot_driver = WebotsController(
          robot_name='my_robot',
          parameters=[
              {'robot_description': robot_description_path}
          ],

          # Every time one resets the simulation the controller is automatically respawned
          respawn=True
      )

      # Starts Webots
      webots = WebotsLauncher(world=PathJoinSubstitution([package_dir, 'worlds', world]))

      return LaunchDescription([
          webots,
          robot_driver
      ])

> On the reset, Webots kills all driver nodes.

Reviewer: ```suggestion
On reset, Webots kills all driver nodes.
```

> Therefore, to start them again after the reset is done, you can set the ``respawn`` property of the driver node to ``True``.

Reviewer: ```suggestion
Therefore, to start them again after reset, you should set the ``respawn`` property of the driver node to ``True``.
```

          executable='spawner',
          arguments=['diffdrive_controller']
      )
      return [
          ros_control_node
      ]

  def generate_launch_description():
      robot_driver = WebotsController(
          robot_name='my_robot',
          parameters=[
              {'robot_description': robot_description_path}
          ],

          # Every time one resets the simulation the controller is automatically respawned
          respawn=True
      )

      # Starts Webots
      webots = WebotsLauncher(world=PathJoinSubstitution([package_dir, 'worlds', world]))

      # Declare the reset handler that respawns nodes when robot_driver exits
      reset_handler = launch.actions.RegisterEventHandler(
          event_handler=launch.event_handlers.OnProcessExit(
              target_action=robot_driver,
              on_exit=get_ros2_control_spawners,
          )
      )

      return LaunchDescription([
          webots,
          robot_driver,
          reset_handler
      ] + get_ros2_control_spawners())

It is not possible to use the ``respawn`` property on the ``ros2_control`` node, as the spawner exits during launch time and not when the simulation is reset.

> Instead we can declare a list of nodes in a function (e.g. ``get_ros2_control_spawners``).

Reviewer: ```suggestion
Instead we should declare a list of nodes in a function (e.g. ``get_ros2_control_spawners``).
```

> This list is started along other nodes when executing the launch file.

Reviewer: ```suggestion
The nodes of this list are started along other nodes when executing the launch file.
```

> It is also declared as action to execute when the ``robot_driver`` node exits, which corresponds to the moment when the simulation is reset in the Webots interface.

Reviewer: ```suggestion
The nodes of this list are also automatically terminated when the ``robot_driver`` node exits, which corresponds to the moment when the simulation is reset from the Webots interface.
```
